mage_integrations/mage_integrations/sources/dynamodb/__init__.py

```diff
@@ -0,0 +1,145 @@
+from botocore.config import Config
+from botocore.exceptions import ClientError
+from datetime import datetime
+from mage_integrations.sources.base import Source, main
+from mage_integrations.sources.catalog import Catalog, CatalogEntry
+from mage_integrations.sources.utils import get_standard_metadata
+from typing import Dict, Generator, List
+import boto3
+# TODO: refactor this common row_to_singer_record function
+# to util file to share between different sources.
+import mage_integrations.sources.mongodb.tap_mongodb.sync_strategies.common as common
+
+REQUIRED_TABLE_CONFIG_KEYS = ['table_name']
+LAST_EVALUATED_TABLE_NAME = 'LastEvaluatedTableName'
+LAST_EVALUATED_KEY = 'LastEvaluatedKey'
+EXCLUSIVE_STAR_KEY = 'ExclusiveStartKey'
+
+
+class DynamoDb(Source):
+    @property
+    def region(self) -> str:
+        return self.config.get('aws_region', 'us-west-2')
+    
+    @property
+    def table_configs(self):
+        """
+        Used for multiple streams. Each table config has the key of table_name.
+        """
+        configs = self.config.get('table_configs', [])
+        configs = [c for c in configs if all(k in c for k in REQUIRED_TABLE_CONFIG_KEYS)]
+        return configs
+
+    def discover_stream(self, client, table_name):
+        '''
+        Read a single table in DynamoDB.
+        '''
+        try:
+            table_info = client.describe_table(TableName=table_name).get('Table', {})
+        except ClientError:
+            self.logger.error(f'Access to table {table_name} was denied, skipping')
+            return None
+
+        key_props = [key_schema.get('AttributeName')
+                     for key_schema in table_info.get('KeySchema', [])]
+        metadata = get_standard_metadata(
+                key_properties=key_props,
+            )
+        catalog_entry = CatalogEntry(
+                key_properties=key_props,
+                metadata=metadata,
+                stream=table_name,
+                tap_stream_id=table_name,
+            )
+        return catalog_entry
+
+    def discover(self, streams: List[str] = None) -> Catalog:
+        '''
+        Read streams in DynamoDB.
+        '''
+        client = self.build_client()
+        self.logger.info(f'Testing streams: {streams}')
+        outputs = []
+        if streams:
+            # Check if streams available
+            for stream in streams:
+                catalog_entry = self.discover_stream(client, stream)
+                if catalog_entry is not None:
+                    outputs.append(catalog_entry)
+        else:
+            for table_config in self.table_configs:
+                catalog_entry = self.discover_stream(client, table_config['table_name'])
+                if catalog_entry is not None:
+                    outputs.append(catalog_entry)
```

Comment:
i don't think we need this. The user will select the streams they want to fetch data from.

Reply:
Done